lib/src/web/database/web_sqlite_database.dart

```diff
@@ -80,38 +80,33 @@ class SqliteDatabaseImpl
     });
   }
 
+  T _runZoned<T>(T Function() callback, {required String debugContext}) {
+    if (Zone.current[this] != null) {
+      throw LockError(
+          'Recursive lock is not allowed. Use `tx.$debugContext` instead of `db.$debugContext`.');
+    }
+    var zone = Zone.current.fork(zoneValues: {this: true});
+    return zone.run(callback);
+  }
+
   @override
   Future<T> readLock<T>(Future<T> Function(SqliteReadContext tx) callback,
       {Duration? lockTimeout, String? debugContext}) async {
     await isInitialized;
-    return _connection.readLock(callback,
-        lockTimeout: lockTimeout, debugContext: debugContext);
+    return _runZoned(() {
+      return _connection.readLock(callback,
+          lockTimeout: lockTimeout, debugContext: debugContext);
+    }, debugContext: debugContext ?? 'execute()');
   }
 
   @override
   Future<T> writeLock<T>(Future<T> Function(SqliteWriteContext tx) callback,
       {Duration? lockTimeout, String? debugContext}) async {
     await isInitialized;
-    return _connection.writeLock(callback,
-        lockTimeout: lockTimeout, debugContext: debugContext);
-  }
-
-  @override
-  Future<T> readTransaction<T>(
```

Comment:
If the semantics you need are "execute if in transaction, otherwise throw", this should be possible with custom requests. `_AsyncSqliteDatabase.handleCustomRequest` has access to the raw database and could implement this request within a single round-trip if necessary.